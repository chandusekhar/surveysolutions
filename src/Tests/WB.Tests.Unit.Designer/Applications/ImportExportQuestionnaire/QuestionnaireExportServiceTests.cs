using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using ApprovalTests.Reporters.TestFrameworks;
using AutoMapper;
using Main.Core.Entities.SubEntities;
using Microsoft.Extensions.Logging;
using Moq;
using NUnit.Framework;
using WB.Core.BoundedContexts.Designer.ImportExport;
using WB.Core.BoundedContexts.Designer.Services;
using WB.Core.BoundedContexts.Designer.Views.Questionnaire.Edit;
using WB.Infrastructure.Native.Files.Implementation.FileSystem;
using WB.Tests.Abc;
using WB.UI.Designer.Code.ImportExport;

namespace WB.Tests.Unit.Designer.Applications.ImportExportQuestionnaire
{
    [UseApprovalSubdirectory("QuestionnaireExportServiceTests-approved")]
    [IgnoreLineEndings(true)]
    [UseReporter(typeof(DiffReporter), typeof(NUnitReporter))]
    [TestOf(typeof(QuestionnaireExportService))]
    public class QuestionnaireExportServiceTests
    {
        [Test]
        public void when_export_only_questionnaire_should_generate_valid_zip()
        {
            var questionnaireDocument = Create.QuestionnaireDocumentWithCoverPage(id: Id.g1, coverId: Id.g2, 
                children: new []
                {
                    Create.TextQuestion(Id.g3)
                });
            questionnaireDocument.Title = "New Questionnaire";
            var view = new QuestionnaireView(questionnaireDocument, Enumerable.Empty<SharedPersonView>());
            var questionnaireViewFactory = Mock.Of<IQuestionnaireViewFactory>(f =>
                f.Load(It.IsAny<QuestionnaireViewInputModel>()) == view);
            
            var service = CreateQuestionnaireExportService(questionnaireViewFactory);

            using var stream = service.GetBackupQuestionnaire(Id.g1, out var filename);

            ZipArchive zip = new ZipArchive(stream);
            var document = zip.GetEntry("document.json").Open();
            var json = new StreamReader(document, Encoding.UTF8).ReadToEnd();

            Approvals.Verify(json);
        }

        private static QuestionnaireExportService CreateQuestionnaireExportService(
            IQuestionnaireViewFactory questionnaireViewFactory)
        {
            var mapper = new MapperConfiguration(cfg =>
            {
                cfg.AddProfile(new QuestionnaireAutoMapperProfile());
            }).CreateMapper();

            var service = new QuestionnaireExportService(questionnaireViewFactory,
                Mock.Of<IAttachmentService>(),
                Mock.Of<ILookupTableService>(),
                Mock.Of<ITranslationImportExportService>(),
                Mock.Of<ICategoriesImportExportService>(),
                Mock.Of<ILogger<QuestionnaireExportService>>(),
                new FileSystemIOAccessor(),
                Create.InMemoryDbContext(),
                new ImportExportQuestionnaireMapper(mapper),
                new QuestionnaireSerializer());
            return service;
        }
    }
}