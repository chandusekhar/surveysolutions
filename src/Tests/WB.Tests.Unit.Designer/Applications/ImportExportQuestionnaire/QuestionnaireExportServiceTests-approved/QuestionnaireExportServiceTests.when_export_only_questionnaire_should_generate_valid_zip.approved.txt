{
  "Id": "11111111-1111-1111-1111-111111111111",
  "CoverPage": {
    "Title": "Chapter",
    "Children": [
      {
        "VariableName": "text_question_333333333333333333",
        "QuestionText": "Question Text test",
        "Id": "33333333-3333-3333-3333-333333333333",
        "Type": "TextQuestion"
      }
    ],
    "Id": "22222222-2222-2222-2222-222222222222",
    "Type": "CoverPage"
  },
  "Title": "New Questionnaire",
  "VariableName": "Q"
}